apps/prairielearn/assets/scripts/examTimeLimitCountdown.ts

```diff
@@ -36,9 +72,28 @@ onDocumentReady(() => {
         form.submit();
       }
     },
-    onServerUpdateFail: () => {
+    onRemainingTime: {
+      60000: () => {
+        if (timeLimitData.showsTimeoutWarning) {
+          showWarningPopup(
+            'examTimeout',
+            'Your exam is ending soon. Please finish and submit your work.',
+          );
+        }
+      },
+    },
+    onServerUpdateFail: (lastRemainingMS: number) => {
       // On time limit fail, reload the page
-      window.location.reload();
+      if (timeLimitData.reloadOnFail) {
+        window.location.reload();
+      } else {
+        if (lastRemainingMS > 0) {
```

Comment:
```suggestion
        // Once the exam access period finishes, we expect updates of the
        // remaining time to fail, so we'll only show the warning popup if
        // time hasn't yet expired.
        if (lastRemainingMS > 0) {
```

AFAICT, `lastRemainingMS` should now be called `remainingMS` since it's computed the instant that this function is called and isn't based on any previous iteration/value/etc., is that correct?

Reply:
Both good points.